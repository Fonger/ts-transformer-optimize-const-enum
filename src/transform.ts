import ts from 'typescript';
import { evaluate, hasModifier } from './utils';

export default function(program: ts.Program, pluginOptions?: unknown) {
  return (ctx: ts.TransformationContext) => {
    return (sourceFile: ts.SourceFile) => {
      const ambient = sourceFile.isDeclarationFile;

      return ts.visitEachChild(sourceFile, visitor, ctx);

      function visitor(node: ts.Node): ts.Node {
        if (!ts.isEnumDeclaration(node)) {
          return ts.visitEachChild(node, visitor, ctx);
        }
        if (
          !hasModifier(node, ts.SyntaxKind.ExportKeyword)
          || !hasModifier(node, ts.SyntaxKind.ConstKeyword)
        ) {
          return node;
        }

        if (ambient) {
          return ts.visitEachChild(node, stripConstKeyword, ctx);
        }

        return transformEnum(node);
      }
    };
  };

  function stripConstKeyword(node: ts.Node) {
    return node.kind === ts.SyntaxKind.ConstKeyword ? undefined : node;
  }

  function transformEnum(node: ts.EnumDeclaration) {
    const members = node.members;
    const known = new Map<string, number | string>();
    const ast = [];
    let lastValue: string | number = -1;

    for (const member of members) {
      if (!ts.isIdentifier(member.name)) {
        throw new Error('The name of enum member must be an identifier');
      }
      const name = member.name.getText();
      let value: string | number;

      if (member.initializer) {
        value = evaluate(member.initializer, known);
        let valueExpr: ts.Expression;
        if (typeof value === 'number') {
          valueExpr = ts.factory.createNumericLiteral(value);
          lastValue = value;
        } else if (typeof value === 'string') {
          valueExpr = ts.factory.createStringLiteral(value);
          lastValue = value;
        } else throw new Error('Unsupported member value');

        ast.push(ts.factory.createPropertyAssignment(name, valueExpr));
      } else {
        if (typeof lastValue === 'string') {
          throw new Error('Invalid enum');
        }
        /**
         * last known value+1
         * enum MyEnum {
         *   A = 5,
         *   B, // implicit 6
         *   C, // implicit 7,
         *   D = 1000,
         *   E, // implicit 1001
         * }
         */
        value = ++lastValue;
        ast.push(
          ts.factory.createPropertyAssignment(
            name,
            ts.factory.createNumericLiteral(value),
          ),
        );
      }
      known.set(name, value);
    }

    return ts.factory.createVariableStatement(
      [
        ts.factory.createModifier(ts.SyntaxKind.ExportKeyword),
        ts.factory.createModifier(ts.SyntaxKind.ConstKeyword),
      ],
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            node.name,
            undefined,
            undefined,
            ts.factory.createAsExpression(
              ts.factory.createObjectLiteralExpression(ast, true),
              ts.factory.createTypeReferenceNode(
                ts.factory.createIdentifier('const'),
                undefined,
              ),
            ),
          ),
        ],
        ts.NodeFlags.Const,
      ),
    );
  }
}
