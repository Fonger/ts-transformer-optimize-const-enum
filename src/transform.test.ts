import ts from 'typescript';
import transform from './transform';
describe('transform export const enum into object literal', () => {
  const output = compile('test-case/const-enum.ts');

  for (const [fileName, content] of Object.entries(output)) {
    it(fileName, () => {
      expect(content).toMatchSnapshot();
    });
  }
});

type Output = { [fileName: string]: string };
function compile(fileName: string): Output {
  const program = ts.createProgram({
    rootNames: [fileName],
    options: {
      target: ts.ScriptTarget.ES2015,
      module: ts.ModuleKind.ES2015,
      preserveConstEnums: true,
      declaration: true,
      sourceMap: true,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
    },
  });

  const transformer = transform(program);
  const output: Output = {};

  program.emit(
    undefined,
    (fileName, data) => {
      output[fileName] = (output[fileName] || '') + data;
    },
    undefined,
    false,
    {
      before: [transformer],
      afterDeclarations: [transformer as any],
    },
  );

  return output;
}
