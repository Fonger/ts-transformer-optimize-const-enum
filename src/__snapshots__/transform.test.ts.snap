// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform export const enum into object literal test-case/const-enum.d.ts 1`] = `
"declare enum SomeEnum {
    A = 0,
    B = 1,
    C = \\"hello\\",
    D = 1000,
    E = 1001
}
declare enum ComputedEnum {
    A = 0,
    B = 1,
    C = 2,
    D = -2,
    E = 4,
    F = -1,
    G = 1228,
    H = 31,
    I = 230,
    J = \\"12\\",
    K = 1
}
export { ComputedEnum as CEnum, SomeEnum };
export declare enum DirectExportEnum {
    HI = \\"HELLO\\",
    WD = \\"WORLD\\"
}
declare enum DefaultExportConstEnum {
    X = 0
}
export default DefaultExportConstEnum;
declare const enum NonExportConstEnum {
    A = 666,
    B = 888
}
declare enum NonExportEnum {
    A = 0,
    B = 1
}
export declare function SomeFunc(inlineSomeEnum: SomeEnum, inlineComputedEnum: ComputedEnum): SomeEnum.D | ComputedEnum | NonExportConstEnum.A | NonExportEnum.B;
"
`;

exports[`transform export const enum into object literal test-case/const-enum.js 1`] = `
"const SomeEnum = {
    A: 0,
    B: 1,
    C: \\"hello\\",
    D: 1000,
    E: 1001
};
const ComputedEnum = {
    A: 0,
    B: 1,
    C: 2,
    D: -2,
    E: 4,
    F: -1,
    G: 1228,
    H: 31,
    I: 230,
    J: \\"12\\",
    K: 1
};
export { ComputedEnum as CEnum, SomeEnum };
export const DirectExportEnum = {
    HI: \\"HELLO\\",
    WD: \\"WORLD\\"
};
const DefaultExportConstEnum = {
    X: 0
};
export default DefaultExportConstEnum;
const Value = 0.5;
var NonExportConstEnum;
(function (NonExportConstEnum) {
    NonExportConstEnum[NonExportConstEnum[\\"A\\"] = 666] = \\"A\\";
    NonExportConstEnum[NonExportConstEnum[\\"B\\"] = 888] = \\"B\\";
})(NonExportConstEnum || (NonExportConstEnum = {}));
var NonExportEnum;
(function (NonExportEnum) {
    NonExportEnum[NonExportEnum[\\"A\\"] = 0] = \\"A\\";
    NonExportEnum[NonExportEnum[\\"B\\"] = 1] = \\"B\\";
})(NonExportEnum || (NonExportEnum = {}));
export function SomeFunc(inlineSomeEnum, inlineComputedEnum) {
    if (inlineSomeEnum === 0 /* A */)
        return 1000 /* D */;
    if (inlineComputedEnum === -2 /* D */)
        return 4 /* E */;
    return Math.random() > Value ? 666 /* A */ : NonExportEnum.B;
}
//# sourceMappingURL=const-enum.js.map"
`;

exports[`transform export const enum into object literal test-case/const-enum.js.map 1`] = `"{\\"version\\":3,\\"file\\":\\"const-enum.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"const-enum.ts\\"],\\"names\\":[],\\"mappings\\":\\"AAAA,MAAW,QAAQ;IACjB,IAAC;IACD,IAAC;IACD,UAAW;IACX,OAAQ;IACR,OAAC;EACF;AAED,MAAW,YAAY;IACrB,IAAC;IACD,IAAC;IACD,IAAC;IACD,KAAU;IACV,IAAU;IACV,KAAc;IACd,OAAuB;IACvB,KAA4B;IAC5B,MAAgB;IAChB,OAAa;IACb,IAAoB;EACrB;AACD,OAAO,EAAE,YAAY,IAAI,KAAK,EAAE,QAAQ,EAAE,CAAC;AAE3C,MAAM,OAAY,gBAAgB;IAChC,WAAY;IACZ,WAAY;EACb;AAED,MAAW,sBAAsB;IAC/B,IAAC;EACF;AAED,eAAe,sBAAsB,CAAC;AAEtC,MAAM,KAAK,GAAG,GAAG,CAAC;AAElB,IAAW,kBAGV;AAHD,WAAW,kBAAkB;IAC3B,uDAAO,CAAA;IACP,uDAAO,CAAA;AACT,CAAC,EAHU,kBAAkB,KAAlB,kBAAkB,QAG5B;AAED,IAAK,aAGJ;AAHD,WAAK,aAAa;IAChB,2CAAC,CAAA;IACD,2CAAC,CAAA;AACH,CAAC,EAHI,aAAa,KAAb,aAAa,QAGjB;AAED,MAAM,UAAU,QAAQ,CAAC,cAAwB,EAAE,kBAAgC;IACjF,IAAI,cAAc,cAAe;QAAE,oBAAkB;IACrD,IAAI,kBAAkB,eAAmB;QAAE,iBAAsB;IACjE,OAAO,IAAI,CAAC,MAAM,EAAE,GAAG,KAAK,CAAC,CAAC,aAAsB,CAAC,CAAC,aAAa,CAAC,CAAC,CAAC;AACxE,CAAC\\"}"`;
